File: FE/src/engine/statEvaluator.ts
import { Stats } from '../types/stats.types';
import { ItemTemplate, ItemInstance } from '../types/item.types';
import { ItemRarity, GemType } from '../types/enums';
import { Hero } from '../types/hero.types';

export const RARITY_MULTIPLIERS: Record<ItemRarity, number> = {
  COMMON: 1.0,
  UNCOMMON: 1.25,
  RARE: 1.6,
  EPIC: 2.1,
  LEGENDARY: 3.0,
  MYTHIC: 4.2,
  ANCIENT: 6.0,
};

export const ENHANCE_GROWTH_PER_LEVEL = 0.10; // +10% per +1 level

export function createDefaultStats(): Stats {
  return {
    physAtk: 0,
    magicAtk: 0,
    atkPercent: 0,
    atkSpeed: 1.0,
    critRate: 5.0,
    critDmg: 50.0,
    elemDmgBonus: 0,
    maxHp: 100,
    armor: 0,
    dmgReduction: 0,
    hpRegen: 0,
    lifeSteal: 0,
    physDodge: 0,
    spellEvasion: 0,
    fireRes: 0,
    coldRes: 0,
    lightningRes: 0,
    chaosRes: 0,
    cdr: 0,
    goldBonus: 0,
    chestDropBonus: 0,
    expBonus: 0,
  };
}

export function computeItemStats(template: ItemTemplate, instance: ItemInstance): Stats {
  const base = template.baseStats || {};
  const rarityMult = RARITY_MULTIPLIERS[instance.rarity] || 1.0;
  const iLvlMult = 1.0 + (Math.max(1, instance.itemLevel) - 1) * (template.iLvlScalingFactor || 0.08);
  const enhanceMult = 1.0 + instance.enhanceLevel * ENHANCE_GROWTH_PER_LEVEL;

  const totalScale = rarityMult * iLvlMult * enhanceMult;

  const computed: Stats = {
    physAtk: Math.round((base.physAtk || 0) * totalScale),
    magicAtk: Math.round((base.magicAtk || 0) * totalScale),
    atkPercent: base.atkPercent || 0,
    atkSpeed: base.atkSpeed || 0,
    critRate: base.critRate || 0,
    critDmg: base.critDmg || 0,
    elemDmgBonus: base.elemDmgBonus || 0,
    maxHp: Math.round((base.maxHp || 0) * totalScale),
    armor: Math.round((base.armor || 0) * totalScale),
    dmgReduction: base.dmgReduction || 0,
    hpRegen: Math.round((base.hpRegen || 0) * totalScale),
    lifeSteal: base.lifeSteal || 0,
    physDodge: base.physDodge || 0,
    spellEvasion: base.spellEvasion || 0,
    fireRes: base.fireRes || 0,
    coldRes: base.coldRes || 0,
    lightningRes: base.lightningRes || 0,
    chaosRes: base.chaosRes || 0,
    cdr: base.cdr || 0,
    goldBonus: base.goldBonus || 0,
    chestDropBonus: base.chestDropBonus || 0,
    expBonus: base.expBonus || 0,
  };

  // Add Sub-stats
  if (instance.subStats) {
    addStats(computed, instance.subStats);
  }

  // Add Sockets Gems
  if (instance.sockets) {
    for (const gemId of instance.sockets) {
      const gemStats = evaluateGemStats(gemId);
      addStats(computed, gemStats);
    }
  }

  // Add Blessing
  if (instance.blessingId) {
    const blessingStats = evaluateBlessingStats(instance.blessingId);
    addStats(computed, blessingStats);
  }

  clampStats(computed);
  return computed;
}

export function evaluateGemStats(gemId: string): Partial<Stats> {
  if (!gemId || !gemId.includes('_T')) return {};
  const [type, tierStr] = gemId.split('_T');
  const tier = parseInt(tierStr) || 1;

  switch (type as GemType) {
    case 'RUBY':
      return { physAtk: tier * 15, magicAtk: tier * 15 };
    case 'EMERALD':
      return { critRate: tier * 2, critDmg: tier * 10 };
    case 'SAPPHIRE':
      return { atkSpeed: tier * 0.05, cdr: tier * 3 };
    case 'TOPAZ':
      return { lifeSteal: tier * 1.5, hpRegen: tier * 10 };
    case 'DIAMOND':
      return {
        dmgReduction: tier * 2,
        fireRes: tier * 3,
        coldRes: tier * 3,
        lightningRes: tier * 3,
        chaosRes: tier * 3,
      };
    default:
      return {};
  }
}

export function evaluateBlessingStats(blessingId: string): Partial<Stats> {
  if (blessingId === 'SCROLL_OF_MIGHT') return { atkPercent: 10 };
  if (blessingId === 'SCROLL_OF_AEGIS') return { dmgReduction: 5 };
  if (blessingId === 'SCROLL_OF_FORTUNE') return { goldBonus: 20, chestDropBonus: 10 };
  return {};
}

export function evaluateHeroLiveStats(hero: Hero, templates: Record<string, ItemTemplate>, baseStats?: Stats): Stats {
  const total = baseStats ? { ...baseStats } : createDefaultStats();

  if (hero && hero.equipment) {
    Object.values(hero.equipment).forEach((itemInstance) => {
      if (itemInstance) {
        const tpl = templates[itemInstance.templateId];
        if (tpl) {
          const itemStats = computeItemStats(tpl, itemInstance);
          addStats(total, itemStats);
        }
      }
    });
  }

  clampStats(total);
  return total;
}

export function addStats(target: Stats, addition: Partial<Stats>) {
  for (const key of Object.keys(addition) as (keyof Stats)[]) {
    if (addition[key] !== undefined) {
      target[key] = (target[key] || 0) + addition[key]!;
    }
  }
}

export function clampStats(stats: Stats) {
  stats.critRate = Math.min(100, Math.max(0, stats.critRate));
  stats.critDmg = Math.max(50, stats.critDmg);
  stats.dmgReduction = Math.min(75, Math.max(0, stats.dmgReduction));
  stats.physDodge = Math.min(75, Math.max(0, stats.physDodge));
  stats.spellEvasion = Math.min(75, Math.max(0, stats.spellEvasion));
  stats.fireRes = Math.min(75, Math.max(0, stats.fireRes));
  stats.coldRes = Math.min(75, Math.max(0, stats.coldRes));
  stats.lightningRes = Math.min(75, Math.max(0, stats.lightningRes));
  stats.chaosRes = Math.min(75, Math.max(0, stats.chaosRes));
  stats.cdr = Math.min(50, Math.max(0, stats.cdr));
  stats.atkSpeed = Math.max(0.2, stats.atkSpeed);
}

File: BE/src/main/java/com/worldhero/engine/StatEvaluator.java
package com.worldhero.engine;

import com.worldhero.dto.ItemInstanceDto;
import com.worldhero.dto.ItemTemplateDto;
import com.worldhero.dto.StatsDto;
import com.worldhero.model.enums.GemType;
import org.springframework.stereotype.Component;

import java.util.List;

/**
 * Thuật toán tính toán chỉ số thực tế của trang bị tại Runtime (Template-Instance Evaluator).
 * Cho phép Admin điều chỉnh Template mà không cần sửa DB người chơi.
 */
@Component
public class StatEvaluator {

    // Hệ số tăng chỉ số mỗi cấp cường hóa (+1 đến +15)
    public static final double ENHANCE_GROWTH_PER_LEVEL = 0.10; // +10% base stat per enhance level

    /**
     * Tính toán chỉ số thực tế của 1 món trang bị cụ thể
     */
    public StatsDto computeItemStats(ItemTemplateDto template, ItemInstanceDto instance) {
        if (template == null || instance == null) {
            return new StatsDto();
        }

        StatsDto base = template.getBaseStats();
        if (base == null) {
            base = new StatsDto();
        }

        // 1. Hệ số nhân theo Phẩm Cấp (Rarity Multiplier)
        double rarityMultiplier = instance.getRarity() != null ? instance.getRarity().getStatMultiplier() : 1.0;

        // 2. Hệ số nhân theo Cấp Vật Phẩm (iLvl Scaling)
        double iLvlMultiplier = 1.0 + (Math.max(1, instance.getItemLevel()) - 1) * template.getILvlScalingFactor();

        // 3. Hệ số Cường Hóa (+0 đến +15)
        double enhanceMultiplier = 1.0 + (instance.getEnhanceLevel() * ENHANCE_GROWTH_PER_LEVEL);

        double totalScale = rarityMultiplier * iLvlMultiplier * enhanceMultiplier;

        StatsDto result = StatsDto.builder()
                .physAtk(Math.round(base.getPhysAtk() * totalScale))
                .magicAtk(Math.round(base.getMagicAtk() * totalScale))
                .atkPercent(base.getAtkPercent())
                .atkSpeed(base.getAtkSpeed())
                .critRate(base.getCritRate())
                .critDmg(base.getCritDmg())
                .elemDmgBonus(base.getElemDmgBonus())
                .maxHp(Math.round(base.getMaxHp() * totalScale))
                .armor(Math.round(base.getArmor() * totalScale))
                .dmgReduction(base.getDmgReduction())
                .hpRegen(base.getHpRegen() * totalScale)
                .lifeSteal(base.getLifeSteal())
                .physDodge(base.getPhysDodge())
                .spellEvasion(base.getSpellEvasion())
                .fireRes(base.getFireRes())
                .coldRes(base.getColdRes())
                .lightningRes(base.getLightningRes())
                .chaosRes(base.getChaosRes())
                .cdr(base.getCdr())
                .goldBonus(base.getGoldBonus())
                .chestDropBonus(base.getChestDropBonus())
                .expBonus(base.getExpBonus())
                .build();

        // 4. Cộng Dòng Phụ (Sub-stats)
        if (instance.getSubStats() != null) {
            result.add(instance.getSubStats());
        }

        // 5. Cộng Ngọc Khảm (Gems Inlaid)
        if (instance.getSockets() != null) {
            for (String gemId : instance.getSockets()) {
                StatsDto gemStats = evaluateGemStats(gemId);
                result.add(gemStats);
            }
        }

        // 6. Cộng Giấy Chúc Phúc (Blessing Scroll)
        if (instance.getBlessingId() != null && !instance.getBlessingId().isBlank()) {
            StatsDto blessingStats = evaluateBlessingStats(instance.getBlessingId());
            result.add(blessingStats);
        }

        // 7. Chuẩn hóa các mốc Hard Caps
        result.clamp();
        return result;
    }

    /**
     * Tính chỉ số của từng loại ngọc theo Tier (RUBY_T1 -> RUBY_T5)
     */
    public StatsDto evaluateGemStats(String gemId) {
        if (gemId == null || !gemId.contains("_T")) return new StatsDto();

        String[] parts = gemId.split("_T");
        String gemTypeStr = parts[0];
        int tier = Integer.parseInt(parts[1]); // 1 to 5

        StatsDto gemStats = new StatsDto();
        try {
            GemType gemType = GemType.valueOf(gemTypeStr);
            switch (gemType) {
                case RUBY -> {
                    double flatAtk = tier * 15.0; // Tier 1: +15, Tier 5: +75 flat ATK
                    gemStats.setPhysAtk(flatAtk);
                    gemStats.setMagicAtk(flatAtk);
                }
                case EMERALD -> {
                    gemStats.setCritRate(tier * 2.0); // +2% to +10% Crit
                    gemStats.setCritDmg(tier * 10.0); // +10% to +50% Crit DMG
                }
                case SAPPHIRE -> {
                    gemStats.setAtkSpeed(tier * 0.05); // +0.05 to +0.25 ASPD
                    gemStats.setCdr(tier * 3.0);       // +3% to +15% CDR
                }
                case TOPAZ -> {
                    gemStats.setLifeSteal(tier * 1.5); // +1.5% to +7.5% Lifesteal
                    gemStats.setHpRegen(tier * 10.0);  // +10 to +50 HP/s
                }
                case DIAMOND -> {
                    gemStats.setDmgReduction(tier * 2.0); // +2% to +10% DmgReduction
                    gemStats.setFireRes(tier * 3.0);
                    gemStats.setColdRes(tier * 3.0);
                    gemStats.setLightningRes(tier * 3.0);
                    gemStats.setChaosRes(tier * 3.0);
                }
            }
        } catch (IllegalArgumentException ignored) {}

        return gemStats;
    }

    /**
     * Tính chỉ số của Giấy Chúc Phúc
     */
    public StatsDto evaluateBlessingStats(String blessingId) {
        StatsDto stats = new StatsDto();
        if ("SCROLL_OF_MIGHT".equalsIgnoreCase(blessingId)) {
            stats.setAtkPercent(10.0); // +10% Tổng Dame
        } else if ("SCROLL_OF_AEGIS".equalsIgnoreCase(blessingId)) {
            stats.setDmgReduction(5.0); // +5% DmgReduction
        } else if ("SCROLL_OF_FORTUNE".equalsIgnoreCase(blessingId)) {
            stats.setGoldBonus(20.0);
            stats.setChestDropBonus(10.0);
        }
        return stats;
    }

    /**
     * Tính tổng chỉ số cho toàn bộ Đội hình 3 Heroes (kết hợp trang bị của cả 3 tướng)
     */
    public StatsDto computeTotalPartyStats(List<StatsDto> heroStatsList) {
        StatsDto total = new StatsDto();
        if (heroStatsList != null) {
            for (StatsDto heroStats : heroStatsList) {
                total.add(heroStats);
            }
        }
        total.clamp();
        return total;
    }
}
